libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -881,6 +915,74 @@ string YulUtilFunctions::readFromStorageDynamic(Type const& _type, bool _splitFu
 	});
 }
 
+string YulUtilFunctions::readFromMemory(Type const& _type)
+{
+	return readFromMemoryOrCalldata(_type, false);
+}
+
+string YulUtilFunctions::readFromCalldata(Type const& _type)
+{
+	return readFromMemoryOrCalldata(_type, true);
+}
+
+string YulUtilFunctions::readFromMemoryOrCalldata(Type const& _type, bool _fromCalldata)
+{
+	string functionName =
+		string("read_from_") +
+		(_fromCalldata ? "calldata" : "memory") +
+		"_dynamic_" +
+		_type.identifier();
+
+	// TODO use ABI functions for handling calldata
+	if (_fromCalldata)
+		solAssert(!_type.isDynamicallyEncoded(), "");
+
+	return m_functionCollector->createFunction(functionName, [&] {
+		if (auto refType = dynamic_cast<ReferenceType const*>(&_type))
+		{
+			solAssert(refType->sizeOnStack() == 1, "");
+			solAssert(!_fromCalldata, "");
+
+			return Whiskers(R"(
+				function <functionName>(memPtr) -> value {
+					value := mload(memPtr)
+				}
+			)")
+			("functionName", functionName)
+			.render();
+		}
+
+		solAssert(_type.isValueType(), "");
+
+		if (auto const* funType = dynamic_cast<FunctionType const*>(&_type))
+			if (funType->kind() == FunctionType::Kind::External)
+				return Whiskers(R"(
+					function <functionName>(memPtr) -> addr, selector {
+						let combined := <load>(memPtr)
+						addr, selector := <splitFunction>(combined)
+					}
+				)")
+				("functionName", functionName)
+				("load", _fromCalldata ? "calldataload" : "mload")
+				("splitFunction", splitExternalFunctionIdFunction())
```

Comment:
Needs validation when read from calldata

Reply:
I would put that in the next PR that will focus on calldata

Reply:
Please add a TODO here so that we do not forget it.